goethe/build.gradle

```diff
@@ -2,11 +2,15 @@ apply plugin: 'com.palantir.external-publish-jar'
 apply plugin: 'com.palantir.shadow-jar'
 
 dependencies {
-    api 'com.squareup:javapoet'
+    compileOnlyApi 'com.palantir.javapoet:javapoet'
+    compileOnlyApi 'com.squareup:javapoet'
```

Comment:
If we make these `compileOnly`, then we get compilation errors in consumers:

```
/Volumes/git/example/example/src/main/java/com/palantir/example/Example.java:42: error: cannot access JavaFile
        Goethe.formatAndEmit(file, baseDir);
              ^
  class file for com.squareup.javapoet.JavaFile not found
1 error
```

I believe this is because the methods are overloaded, so the compiler needs to load all overload in order to perform method resolution. If we don't overloading the methods, then we can use `compileOnly` here without causing compilation errors in consumers.

One downside of using `compileOnlyApi` is that these dependencies end up on the runtime classpath of consumers. Unlike `compileOnly` dependencies, which are omitted from the `pom.xml`, `compileOnlyApi` dependencies are included in the `pom.xml` with the `compile` scope, which ultimately end up on the runtime classpath of consumers.

It seemed better to match the existing methods names at the cost of having this additional runtime dependency. `com.squareup:javapoet` is fairly small and has no transitive dependencies. And in the future we can remove support for `com.squareup:javapoet` entirely.

Reply:
It looks like Gradle simply doesn't support the Maven `provided` scope, which has the semantics we want here. The Java plugin only creates 2 consumable configurations - `apiElements` and `runtimeElements`. These correspond to the Maven `compile` and `runtime` scopes, respectively.

https://github.com/gradle/gradle/blob/43fc58d9d2aefd6611f204ac60cbb5be717fed8c/platforms/jvm/plugins-java/src/main/java/org/gradle/api/plugins/JavaPlugin.java#L307-L308

`compileOnlyApi` dependencies end up in the `apiElements` configuration, along with `api` dependencies.

I guess I shouldn't be too surprised, since this is literally what the Gradle docs show.

https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph

This seems like a Gradle deficiency. Although maybe this is intentional/expected? From the [Gradle 6.7 release notes](https://docs.gradle.org/6.7/release-notes.html):

> The behavior of `compileOnlyApi` dependencies is preserved for published libraries when published with [Gradle Module Metadata](https://docs.gradle.org/6.7/userguide/publishing_gradle_module_metadata.html#).